src/libLookup/Lookup.cpp

```diff
@@ -5436,145 +5441,193 @@ void Lookup::RectifyTxnShardMap(const uint32_t oldNumShards,
+      lock_guard<mutex> g(m_mediator.m_mutexDSCommittee);
 
-      if (LOG_PARAMETERS) {
-        LOG_STATE("[TXNPKT][" << m_mediator.m_currentEpochNum << "] Shard=" << i
-                              << " NumTx="
-                              << (GetTxnFromShardMap(i).size() + mp[i].size()));
+      if (m_mediator.m_DSCommittee->empty()) {
+        return;
       }
 
-      if (GetTxnFromShardMap(i).empty() && mp[i].empty()) {
-        LOG_GENERAL(INFO, "No txns to send to shard " << i);
-        continue;
+      // Send to NUM_NODES_TO_SEND_LOOKUP which including DS leader
+      PairOfNode dsLeader;
+      if (Node::GetDSLeader(m_mediator.m_blocklinkchain.GetLatestBlockLink(),
+                            m_mediator.m_dsBlockChain.GetLastBlock(),
+                            *m_mediator.m_DSCommittee, dsLeader)) {
+        toSend.push_back(dsLeader.second);
       }
 
-      result = Messenger::SetNodeForwardTxnBlock(
-          msg, MessageOffset::BODY, m_mediator.m_currentEpochNum,
-          m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum(), i,
-          m_mediator.m_selfKey, GetTxnFromShardMap(i), mp[i]);
-    }
-
-    if (!result) {
-      LOG_EPOCH(WARNING, m_mediator.m_currentEpochNum,
-                "Messenger::SetNodeForwardTxnBlock failed.");
-      LOG_GENERAL(WARNING, "Cannot create packet for " << i << " shard");
-      continue;
-    }
-    vector<Peer> toSend;
-    if (i < numShards) {
-      {
-        lock_guard<mutex> g(m_mediator.m_ds->m_mutexShards);
-        uint16_t lastBlockHash = DataConversion::charArrTo16Bits(
-            m_mediator.m_txBlockChain.GetLastBlock().GetBlockHash().asBytes());
-        uint32_t leader_id = m_mediator.m_node->CalculateShardLeaderFromShard(
-            lastBlockHash, m_mediator.m_ds->m_shards.at(i).size(),
-            m_mediator.m_ds->m_shards.at(i));
-        LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
-                  "Shard leader id " << leader_id);
-
-        auto it = m_mediator.m_ds->m_shards.at(i).begin();
-        // Lookup sends to NUM_NODES_TO_SEND_LOOKUP + Leader
-        unsigned int num_node_to_send = NUM_NODES_TO_SEND_LOOKUP;
-        for (unsigned int j = 0; j < num_node_to_send &&
-                                 it != m_mediator.m_ds->m_shards.at(i).end();
-             j++, it++) {
-          if (distance(m_mediator.m_ds->m_shards.at(i).begin(), it) ==
-              leader_id) {
-            num_node_to_send++;
-          } else {
-            toSend.push_back(std::get<SHARD_NODE_PEER>(*it));
-            LOG_GENERAL(INFO, "Sent to node " << get<SHARD_NODE_PEER>(*it));
-          }
+      for (auto const& i : *m_mediator.m_DSCommittee) {
+        if (toSend.size() < NUM_NODES_TO_SEND_LOOKUP &&
+            i.second != dsLeader.second) {
+          toSend.push_back(i.second);
         }
-        if (m_mediator.m_ds->m_shards.at(i).empty()) {
-          continue;
+
+        if (toSend.size() >= NUM_NODES_TO_SEND_LOOKUP) {
+          return;
```

Comment:
While testing i encountered that this should be `break` instead of `return`.

Reply:
will fix